File: emulators/springcoat-vm/src/main/java/cc/squirreljme/vm/springcoat/MLETerminal.java
// -*- Mode: Java; indent-tabs-mode: t; tab-width: 4 -*-
// ---------------------------------------------------------------------------
// SquirrelJME
//     Copyright (C) Stephanie Gawroriski <xer@multiphasicapps.net>
// ---------------------------------------------------------------------------
// SquirrelJME is under the Mozilla Public License Version 2.0.
// See license.mkd for licensing and copyright information.
// ---------------------------------------------------------------------------

package cc.squirreljme.vm.springcoat;

import cc.squirreljme.emulator.MLECallWouldFail;
import cc.squirreljme.jvm.mle.TerminalShelf;
import cc.squirreljme.jvm.mle.brackets.PipeBracket;
import cc.squirreljme.jvm.mle.constants.PipeErrorType;
import cc.squirreljme.vm.springcoat.brackets.PipeObject;
import cc.squirreljme.vm.springcoat.exceptions.SpringMLECallError;
import java.io.IOException;

/**
 * Functions for {@link MLETerminal}.
 *
 * @since 2020/06/18
 */
public enum MLETerminal
	implements MLEFunction
{
	/** {@link TerminalShelf#close(PipeBracket)}. */
	CLOSE("close:(Lcc/squirreljme/jvm/mle/brackets/PipeBracket;)I")
	{
		/**
		 * {@inheritDoc}
		 * @since 2020/07/06
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			try
			{
				MLEObjects.pipe(__args[0]).pipe.close();
				
				return null;
			}
			catch (IOException|MLECallWouldFail e)
			{
				throw new SpringMLECallError(e.getMessage(), e);
			}
		}
	}, 
	
	/** {@link TerminalShelf#flush(PipeBracket)}. */
	FLUSH("flush:(Lcc/squirreljme/jvm/mle/brackets/PipeBracket;)I")
	{
		/**
		 * {@inheritDoc}
		 * @since 2020/06/18
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			try
			{
				MLEObjects.pipe(__args[0]).pipe.flush();
				
				return PipeErrorType.NO_ERROR;
			}
			catch (IOException|MLECallWouldFail e)
			{
				throw new SpringMLECallError(e.getMessage(), e);
			}
		}
	},
	
	/** {@link TerminalShelf#fromStandard(int)}. */
	FROM_STANDARD("fromStandard:(I)" +
		"Lcc/squirreljme/jvm/mle/brackets/PipeBracket;")
	{
		/**
		 * {@inheritDoc}
		 * @since 2022/03/19
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			try
			{
				return new PipeObject(__thread.machine,
					__thread.machine.terminalPipes.mleGet((int)__args[0]));
			}
			catch (MLECallWouldFail e)
			{
				throw new SpringMLECallError(e);
			}
		}
	}, 
	
	/** {@link TerminalShelf#write(PipeBracket, int)}. */
	WRITE_BYTE("write:(Lcc/squirreljme/jvm/mle/brackets/PipeBracket;I)I")
	{
		/**
		 * {@inheritDoc}
		 * @since 2020/06/18
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			try
			{
				MLEObjects.pipe(__args[0]).pipe
					.write((int)__args[1]);
				
				return 1;
			}
			catch (IOException|MLECallWouldFail e)
			{
				throw new SpringMLECallError(e.getMessage(), e);
			}
		}
	},
	
	/** {@link TerminalShelf#write(PipeBracket, byte[], int, int)}. */
	WRITE_BYTES("write:(Lcc/squirreljme/jvm/mle/brackets/PipeBracket;" +
		"[BII)I")
	{
		/**
		 * {@inheritDoc}
		 * @since 2020/06/18
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			try
			{
				if (!(__args[1] instanceof SpringArrayObjectByte))
					throw new SpringMLECallError("Not a byte array.");
				
				SpringArrayObjectByte buf = (SpringArrayObjectByte)__args[1];
				int off = (int)__args[2];
				int len = (int)__args[3];
				
				MLEObjects.pipe(__args[0]).pipe
					.write(buf.array(), off, len);
				return len;
			}
			catch (IOException|MLECallWouldFail e)
			{
				throw new SpringMLECallError(e.getMessage(), e);
			}
		}
	},
	
	/* End. */
	;
	
	/** The dispatch key. */
	protected final String key;
	
	/**
	 * Initializes the dispatcher info.
	 *
	 * @param __key The key.
	 * @throws NullPointerException On null arguments.
	 * @since 2020/06/18
	 */
	MLETerminal(String __key)
		throws NullPointerException
	{
		if (__key == null)
			throw new NullPointerException("NARG");
		
		this.key = __key;
	}
	
	/**
	 * {@inheritDoc}
	 * @since 2020/06/18
	 */
	@Override
	public String key()
	{
		return this.key;
	}
	
}

File: emulators/emulator-base-native/c/mle_terminal.c
/* -*- Mode: C++; indent-tabs-mode: t; tab-width: 4 -*-
// ---------------------------------------------------------------------------
// SquirrelJME
//	 Copyright (C) Stephanie Gawroriski <xer@multiphasicapps.net>
// ---------------------------------------------------------------------------
// SquirrelJME is under the Mozilla Public License Version 2.0.
// See license.mkd for licensing and copyright information.
// -------------------------------------------------------------------------*/

#include <stdio.h>

/* //// MLE /// */
#define mleGroupId TerminalShelf
#define mleShelfClass "cc/squirreljme/jvm/mle/TerminalShelf"
#define mleProxyTarget "cc/squirreljme/emulator/EmulatedTerminalShelf"
#include "squirreljmeMle.h"
/* //////////// */

#include "squirreljme.h"

#define MLE_DESC_available DESC_METHOD(DESC_INT, \
	DESC_PIPE)
MLE_FUNC_PROXY_STATIC(jint, available)

#define MLE_DESC_close DESC_METHOD(DESC_INT, \
	DESC_PIPE)
MLE_FUNC_PROXY_STATIC(jint, close)

#define MLE_DESC_flush DESC_METHOD(DESC_INT, \
	DESC_PIPE)
MLE_FUNC_PROXY_STATIC(jint, flush)

#define MLE_DESC_fromStandard DESC_METHOD(DESC_PIPE, \
	DESC_INT)
MLE_FUNC_PROXY_STATIC(jobject, fromStandard)

#define MLE_DESC_read_single DESC_METHOD(DESC_INT, \
	DESC_PIPE)
MLE_FUNC_PROXY_STATIC_ALT(jint, read, single)

#define MLE_DESC_read_multi DESC_METHOD(DESC_INT, \
	DESC_PIPE DESC_ARRAY(DESC_BYTE) DESC_INT DESC_INT)
MLE_FUNC_PROXY_STATIC_ALT(jint, read, multi)

#define MLE_DESC_write_single DESC_METHOD(DESC_INT, \
	DESC_PIPE DESC_INT)
MLE_FUNC_PROXY_STATIC_ALT(jint, write, single)

#define MLE_DESC_write_multi DESC_METHOD(DESC_INT, \
	DESC_PIPE DESC_ARRAY(DESC_BYTE) DESC_INT DESC_INT)
MLE_FUNC_PROXY_STATIC_ALT(jint, write, multi)

MLE_LIST_BEGIN()
	MLE_LIST_ITEM(available),
	MLE_LIST_ITEM(close),
	MLE_LIST_ITEM(flush),
	MLE_LIST_ITEM(fromStandard),
	MLE_LIST_ITEM_ALT(read, single),
	MLE_LIST_ITEM_ALT(read, multi),
	MLE_LIST_ITEM_ALT(write, single),
	MLE_LIST_ITEM_ALT(write, multi),
MLE_LIST_END()
